# Описание


Коронавирус застал мир врасплох, изменив привычный порядок вещей. В
свободное время жители городов больше не выходят на улицу, не посещают кафе
и торговые центры. Зато стало больше времени для книг. Это заметили
стартаперы — и бросились создавать приложения для тех, кто любит читать.
Ваша компания решила быть на волне и купила крупный сервис для чтения книг
по подписке. Ваша первая задача как аналитика — проанализировать базу
данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские
обзоры книг. Эти данные помогут сформулировать ценностное предложение для
нового продукта.

# Описание данных

**Таблица** `books`

Содержит данные о книгах:
- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица** `authors`

Содержит данные об авторах:
- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица** `publishers`

Содержит данные об издательствах:
- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица ratings**

Содержит данные о пользовательских оценках книг:
- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица reviews**

Содержит данные о пользовательских обзорах на книги:
- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

# Задания

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще - 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.


# Как выполнить задание?

- Опишите цели исследования;
- Исследуйте таблицы — выведите первые строки;
- Сделайте по одному SQL-запросу для решения каждого задания;
- Выведите результат каждого запроса в тетрадке;

## Откройте файл с данными и изучите общую информацию

In [1]:
import pandas as pd
from sqlalchemy import create_engine

In [2]:
db_config = {'user': 'praktikum_student', # имя пользователя
             'pwd': 'Sdf4$2;d-d30pp', # пароль
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432, # порт подключения
             'db': 'data-analyst-final-project-db'} # название базы данных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                         db_config['pwd'],
                                                         db_config['host'],
                                                         db_config['port'],
                                                         db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

In [3]:
query = '''
            SELECT * FROM books
        '''

books = pd.io.sql.read_sql(query, con = engine)
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           1000 non-null   int64 
 1   author_id         1000 non-null   int64 
 2   title             1000 non-null   object
 3   num_pages         1000 non-null   int64 
 4   publication_date  1000 non-null   object
 5   publisher_id      1000 non-null   int64 
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


In [4]:
books.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [5]:
query = '''
            SELECT * FROM authors
        '''

authors = pd.io.sql.read_sql(query, con = engine)
authors.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


In [6]:
authors.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [7]:
query = '''
            SELECT * FROM publishers
        '''

publishers = pd.io.sql.read_sql(query, con = engine)
publishers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   publisher_id  340 non-null    int64 
 1   publisher     340 non-null    object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


In [8]:
publishers.head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [9]:
query = '''
            SELECT * FROM ratings
        '''

ratings = pd.io.sql.read_sql(query, con = engine)
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


In [10]:
ratings.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [11]:
query = '''
            SELECT * FROM reviews
        '''

reviews = pd.io.sql.read_sql(query, con = engine)
reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


In [12]:
reviews.head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

С данными все в порядке, ошибок в типах данных нет

## Подготовьте данные 

### Обработка дубликатов

In [13]:
books.duplicated().sum()

0

In [14]:
authors.duplicated().sum()

0

In [15]:
publishers.duplicated().sum()

0

In [16]:
ratings.duplicated().sum()

0

In [17]:
reviews.duplicated().sum()

0

Дубликатов в наших датасетах нет

In [18]:
books.isna().sum()

book_id             0
author_id           0
title               0
num_pages           0
publication_date    0
publisher_id        0
dtype: int64

In [19]:
authors.isna().sum()

author_id    0
author       0
dtype: int64

In [20]:
publishers.isna().sum()

publisher_id    0
publisher       0
dtype: int64

In [21]:
ratings.isna().sum()

rating_id    0
book_id      0
username     0
rating       0
dtype: int64

In [22]:
reviews.isna().sum()

review_id    0
book_id      0
username     0
text         0
dtype: int64

**Вывод по шаг 2**:
Дубликатов и пропусков нет

## Задание

### Посчитайте, сколько книг вышло после 1 января 2000 года;

In [23]:
query = '''
SELECT 
    COUNT(book_id) 
FROM 
    books
WHERE 
    publication_date > '2000-01-01'
'''

number_books = pd.io.sql.read_sql(query, con = engine)
print('C 1 января 2000 года выпущено {} книг'.format(number_books.loc[:, 'count'][0]))

C 1 января 2000 года выпущено 819 книг


### Для каждой книги посчитайте количество обзоров и среднюю оценку;

In [29]:
query = '''
SELECT 
    book.title,
    CASE
        WHEN review.reviews_num IS NULL THEN 0
        ELSE review.reviews_num
    END AS reviews_num,
    CASE
        WHEN rating.avg_rating IS NULL THEN 0
        ELSE rating.avg_rating
    END AS avg_rating
FROM
    books book
LEFT JOIN 
    (SELECT
        book_id,
        COUNT(review_id) reviews_num
    FROM 
        reviews
    GROUP BY 
        book_id
    ) AS review ON book.book_id = review.book_id
LEFT JOIN 
    (SELECT
        book_id,
        AVG(rating :: real) avg_rating
    FROM 
        ratings
    GROUP BY 
        book_id
    ) AS rating ON book.book_id = rating.book_id    
ORDER BY
    reviews_num DESC
LIMIT 10
'''

book_count_mean = pd.io.sql.read_sql(query, con = engine)
book_count_mean.sort_values(by='avg_rating', ascending=False)

title  reviews_num  avg_rating
3                                   The Glass Castle            6    4.206897
4                The Hobbit  or There and Back Again            6    4.125000
6                          Outlander (Outlander  #1)            6    4.125000
7  The Curious Incident of the Dog in the Night-Time            6    4.081081
8                                Water for Elephants            6    3.977273
2             The Da Vinci Code (Robert Langdon  #2)            6    3.830508
1                                      The Alchemist            6    3.789474
5                                           The Road            6    3.772727
0                            Twilight (Twilight  #1)            7    3.662500
9                                    Eat  Pray  Love            6    3.395833

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц

In [25]:
publishers_books = '''SELECT
        publishers.publisher AS publisher,
        COUNT(books.book_id) AS total_books
        FROM
            books
        LEFT JOIN publishers ON 
            publishers.publisher_id = books.publisher_id
        GROUP BY 
            publishers.publisher
        HAVING
            SUM(books.num_pages) > 50
        ORDER BY 
            total_books DESC
        LIMIT 10
'''
publishers_books = pd.io.sql.read_sql(publishers_books, con = engine)
publishers_books

publisher  total_books
0              Penguin Books           42
1                    Vintage           31
2   Grand Central Publishing           25
3           Penguin Classics           24
4           Ballantine Books           19
5                     Bantam           19
6                    Berkley           17
7         St. Martin's Press           14
8              Berkley Books           14
9  William Morrow Paperbacks           13

### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

In [31]:
author_rating = '''SELECT
    authors.author,
    AVG(filtered.average_rating) AS avg_rating
    FROM
        (SELECT
            book_id AS book_id,
            COUNT(rating_id) AS rating_counts,
            AVG(rating) AS average_rating
        FROM
            ratings
        GROUP BY book_id
        HAVING COUNT(rating_id) >= 50) 
        AS filtered
    INNER JOIN books ON books.book_id = filtered.book_id
    INNER JOIN authors ON authors.author_id = books.author_id
    GROUP BY authors.author
    ORDER BY avg_rating DESC
    LIMIT 10
'''
author_rating = pd.io.sql.read_sql(author_rating, con = engine)
author_rating

author  avg_rating
0                         J.K. Rowling/Mary GrandPré    4.283844
1                  Markus Zusak/Cao Xuân Việt Khương    4.264151
2                                     J.R.R. Tolkien    4.258446
3                                  Louisa May Alcott    4.192308
4                                       Rick Riordan    4.080645
5                                    William Golding    3.901408
6                                      J.D. Salinger    3.825581
7           Paulo Coelho/Alan R. Clarke/Özdemir İnce    3.789474
8  William Shakespeare/Paul Werstine/Barbara A. M...    3.787879
9                                          Dan Brown    3.754540

### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [27]:
query = '''
SELECT
    AVG(reviews_num)
FROM
    (SELECT
        review.username,
        COUNT(*) reviews_num
    FROM
        reviews review
    INNER JOIN 
        (SELECT 
            username
        FROM
            ratings
        GROUP BY
            username 
        HAVING 
            COUNT(rating_id) > 50
        ) AS users ON users.username = review.username
    GROUP BY
        review.username
    ) rev
'''

average_number_reviews = pd.io.sql.read_sql(query, con = engine)
average_number_reviews

avg
0  24.333333

**Выводы**

- C 1 января 2000 года выпущено 819 книг
- Самое большое количество обзоров у книги Twilight (Twilight #1), а самый большой рейтинг у книги The Glass Castle
- Издательство, которое выпустило наибольшее число книг толще 50 страниц Penguin Books - 42 книги
- Автор с самой высокой средней оценкой книг, учитывая книги более 50 страниц - J.K. Rowling/Mary GrandPré. Рейтинг - 4.28.
- Среднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24.33.